// eslint-nibble/lib/config/options.js
/**
 * @fileoverview Options configuration for optionator.
 * @author Ian VanSchooten
 */
'use strict';

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _optionator = require('optionator');

var _optionator2 = _interopRequireDefault(_optionator);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//------------------------------------------------------------------------------
// Initialization and Public Interface
//------------------------------------------------------------------------------


exports.default = (0, _optionator2.default)({
  prepend: 'usage: eslint-nibble [file.js] [dir]',
  options: [{
    heading: 'Options'
  }, {
    option: 'help',
    alias: 'h',
    type: 'Boolean',
    description: 'Show help'
  }, {
    option: 'version',
    alias: 'v',
    type: 'Boolean',
    description: 'Outputs the version number'
  }, {
    option: 'ext',
    type: '[String]',
    default: '.js',
    description: 'Specify JavaScript file extensions'
  }, {
    option: 'config',
    alias: 'c',
    type: 'path::String',
    description: 'Use configuration from this file or shareable config'
  }, {
    option: 'cache',
    type: 'Boolean',
    default: 'false',
    description: 'Only check changed files'
  }, {
    option: 'cache-location',
    type: 'path::String',
    description: 'Path to the cache file or directory'
  }, {
    option: 'rule',
    type: '[String]',
    concatRepeatedArrays: true,
    description: 'Only show results for specified rule(s)'
  }, {
    option: 'warnings',
    type: 'Boolean',
    default: 'true',
    description: 'include warning results'
  }]
});
